### PR TITLE
cdn: URL signing example

### PR DESCRIPTION
**Pull request by DanSanche**
I added some samples for. Cloud CDN URL signing. I believe we don't want to merge this until the feature is actually ready to launch

this is my first submission to the sample repo, so let me know if I'm doing anything weird.

### CODE REVIEW

**Comment by DanSanche** [comment]
@frankyn Would you mind double-checking to make sure I'm doing the maven stuff properly?

@ahmetb

**Comment by ahmetb** [comment]
Reviewers: please review, but **DO NOT MERGE** until we're ready to launch.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * Samples to create a signed URL for a Cloud CDN endpoint
+ */
+public class SignedUrls {
+
+  // [START signUrl]
+  /**
+   * Creates a signed URL for a Cloud CDN endpoint with the given key
+   * should pass in a properly formatted URL
+   * will work:
```
Comment (ahmetb) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: you can remove the following examples and just say

`@param url the the Cloud CDN endpoint to sign, make sure there's a forward slash (/) after the hostname.`

File: cdn/signed-urls/src/main/resources/my-key [hunk 2]
```diff
@@ -0,0 +1 @@
+nZtRohdNF9m3cKM24IcK4w==
```
Comment (ahmetb) [inline cdn/signed-urls/src/main/resources/my-key]: do not check this in

File: cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlsTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Test SignedUrls samples
+ */
+@RunWith(JUnit4.class)
+@SuppressWarnings("checkstyle:abbreviationaswordinname")
+public class SignedUrlsTest {
+  private static long TIMESTAMP = 1518135754;
+  private static Date EXPIRATION = new Date(TIMESTAMP * 1000);
+  private static byte[] KEY_BYTES = Base64.getUrlDecoder().decode("aaaaaaaaaaaaaaaaaaaaaa==");
+  private static String KEY_NAME = "my-key";
+  private static String BASE_URL = "https://www.google.com/";
```
Comment (ahmetb) [inline cdn/signed-urls/src/test/java/com/google/cdn/SignedUrlsTest.java]: use example.com for tests

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    cal.add(Calendar.DATE, 1);
+    Date tomorrow = cal.getTime();
+
+    //read the key as a base 64 url-safe encoded string
+    final String keyPath = "./src/main/resources/my-key";
+    String base64String = new String(Files.readAllBytes(Paths.get(keyPath)));
+    //turn the key string into a byte array
+    byte[] keyBytes = Base64.getUrlDecoder().decode(base64String);
+
+    String result = signUrl("http://35.186.234.33/index.html", keyBytes, "my-key", tomorrow);
```
Comment (ahmetb) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: use http://example.com/ here and `YOUR-KEY-NAME` (uppercase indicates people should change it)

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  // [END signUrl]
+
+  public static void main(String[] args) throws Exception {
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(new Date());
+    cal.add(Calendar.DATE, 1);
+    Date tomorrow = cal.getTime();
+
+    //read the key as a base 64 url-safe encoded string
+    final String keyPath = "./src/main/resources/my-key";
```
Comment (ahmetb) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: use `/path/to/your/key`

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    String base64String = new String(Files.readAllBytes(Paths.get(keyPath)));
+    //turn the key string into a byte array
```
Comment (ahmetb) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: redundant comment

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    String encoded = SignedUrls.getSignature(key, urlToSign);
+    return urlToSign + "&Signature=" + encoded;
+  }
+
+  public static String getSignature(byte[] privateKey, String input)
+      throws InvalidKeyException, NoSuchAlgorithmException {
+
+    final String algorithm = "HmacSHA1";
+    Key key = new SecretKeySpec(privateKey, 0, privateKey.length, algorithm);
```
Comment (frankyn) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: What does this 0 mean?

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+   *   https://www.google.com/?foo=test&bar=test
+   *   https://www.google.com/foo
+   *
+   * won't work:
+   *    https://www.google.com
+   *    https://www.google.com?test
+   *    www.google.com
+   *
+   * @param url the Cloud CDN endpoint to sign
+   * @param key encoded as a 16-byte array
```
Comment (ahmetb) [inline cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: `encoded as a 16-byte array` --> `url signing key uploaded to the backend service/bucket, as a 16-byte array`

"16-byte array" is not an encoding :)

File: pom.xml [hunk 9]
```diff
@@ -47,6 +47,7 @@
 
         <module>cloud-tasks</module>
         <module>compute</module>
+        <module>cdn/signed-urls</module>
```
Comment (frankyn) [inline pom.xml]: LGTM and tests are run.

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java (hunk 7 above)
Comment (DanSanche) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: It looks like the function can take an offset for when there is garbage data before the key. Do you think I should give the 0 a name?

File: pom.xml (hunk 9 above)
Comment (DanSanche) [reply pom.xml]: Thanks for checking!

File: cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java (hunk 8 above)
Comment (DanSanche) [reply cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java]: I've addressed the issues you mentioned, let me know what you think

**Comment by ahmetb** [comment]
@DanSanche can you rebase to address the merge conflict?

This is now OK to merge from logistics perspective. If maintainers can take a look and merge, I appreciate.

**Comment by ahmetb** [comment]
**Repo owners:** It looks like there's a requirement like "PR branches must include all commits from the master" in this repo. Do we want to retain this policy? Often times, this will cause inconvenience to the contributors as they have to rebase multiple times, even though there are no merge-conflicts.

If someone can review/LGTM, that'll be appreciated.

File: cdn/signed-urls/pom.xml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    <groupId>com.google.cloud.samples</groupId>
+    <artifactId>shared-configuration</artifactId>
+    <version>1.0.8</version>
+  </parent>
+
+  <name>signedurls</name>
+  <url>http://maven.apache.org</url>
+
+  <properties>
+    <maven.compiler.target>1.7</maven.compiler.target>
```
Comment (lesv) [inline cdn/signed-urls/pom.xml]: 1.8

**Comment by DanSanche** [comment]
@lesv I added the README and fixed the version number

**Comment by lesv** [comment]
I know all the contributors have signed the CLA

**Pull request merged**